workflow/executor/kubelet/kubelet.go

```diff
@@ -0,0 +1,102 @@
+package kubelet
+
+import (
+	"compress/gzip"
+	"os"
+	"syscall"
+	"time"
+
+	"github.com/argoproj/argo/errors"
+	log "github.com/sirupsen/logrus"
+)
+
+// killGracePeriod is the time in seconds after sending SIGTERM before
+// forcefully killing the sidecar with SIGKILL (value matches k8s)
+const killGracePeriod = 30
+
+type KubeletExecutor struct {
+	cli *kubeletClient
+}
+
+func NewKubeletExecutor() (*KubeletExecutor, error) {
+	log.Infof("Creating a kubelet executor")
+	cli, err := newKubeletClient()
+	if err != nil {
+		return nil, errors.InternalWrapError(err)
+	}
+	return &KubeletExecutor{
+		cli: cli,
+	}, nil
+}
+
+func (k *KubeletExecutor) GetFileContents(containerID string, sourcePath string) (string, error) {
+	b, err := k.cli.GetFileContents(containerID, sourcePath)
+	if err != nil {
+		return "", err
+	}
+	return b.String(), nil
+}
+
+func (k *KubeletExecutor) CopyFile(containerID string, sourcePath string, destPath string) error {
+	log.Infof("Archiving %s:%s to %s", containerID, sourcePath, destPath)
+	b, err := k.cli.GetFileContents(containerID, sourcePath)
```

Comment:
@jessesuen I thought we are only getting a regular file here. Of course I can change it to use `tar` instead.

Reply:
Directories can be artifacts, so we do need to cover that case

Reply:
I'd like to get your opinion on:
`GetFileContents` should only work on a regular file? If it is, I think it's still adapted to use `cat` there instead of tar / untar its content.

In that case, what do you think to change the prototype from
```go
CopyFile(containerID string, sourcePath string, destPath string)
```
to:
```go
CopyFile(containerID string, sourcePath string, destPath string, dir bool)
```

And just switch between `cat` and `tar` ?

Reply:
> GetFileContents should only work on a regular file

Correct, GetFileContents only needs to work with regular files. You can see the docker implementation doing the following to get the file contents as a string:

`docker cp -a %s:%s - | tar -ax -O` 

The kubelet executor could just use `cat` to achieve the same effect to avoid the tar/untar.

> And just switch between cat and tar ?

Unlike GetFileContents, CopyFile needs to work with files and directories, so `tar` seems the most appropriate and doesn't require a switch. If you see our docker implementation (which handles both directories and files), it accomplishes this using:

`docker cp -a <containerid>:/path/to/file/or/dir -`

which sends a tar stream of the file/directory to stdout.